### PR TITLE
Nexus: Delete state machine on terminal state -- Part 2

### PR DESCRIPTION
**Pull request by justinp-tt**
## What changed?
Tree output compaction & Operation Log Refactor

This PR refactors the hierarchical state machine's operation tracking to better handle node deletions and provide a clearer interface for tracking state changes.

### Key Changes
- Introduced new `Operation` interface and concrete types:
  - `DeleteOperation`: Represents node deletions
  - `TransitionOperation`: Represents state transitions
  - `OperationLog`: Ordered sequence of operations that maintain chronological order
- Improved deletion handling:
  - Filters out transitions for deleted nodes while preserving deletion records
  - Maintains operation ordering during compaction
  - Properly handles deletion of nested hierarchies
- Simplified output retrieval:
  - Removed `PathAndOutputs` in favor of `Operation` interface
  - Operations carry their own path information
  - Consolidated filtering logic in `Compact()`

### Behavioral Changes
When a node is deleted:
- Only the deletion operation is preserved in the log
- All transitions on that node and its descendants are filtered out
- Operations on unrelated paths remain intact and maintain their order
- Re-adding deleted nodes is not supported

### Areas for Review Focus
1. **Deletion Semantics**: Verify that the deletion behavior (recording, filtering, visibility) makes sense
2. **Operation Interface**: Review if the Operation interface and concrete types provide the right abstraction level
3. **Compaction Logic**: Review the filtering and ordering guarantees
4. **Edge Cases**:
   - Path prefix handling (e.g., "node1" vs "node10")
   - Deep hierarchies with multiple deletion points
   - Operations ordering when interleaved with deletions

### Tests Added/Modified
- `TestNode_DeleteChild`: Basic deletion and operation visibility
- `TestNode_PreservesUnrelatedOperations`: Operation preservation across deletions
- `TestNode_PathPrefixEdgeCases`: Path matching and ambiguity handling
- `TestNode_ComplexHierarchicalDeletions`: Deep hierarchy deletion behavior
- `TestNode_CompactionOrderPreservation`: Operation ordering guarantees
- Additional tests for edge cases and operation ordering

## Why?
Prerequisite to supporting state machine deletion on terminal state. This refactor provides a more robust foundation for tracking state machine operations and handling deletions consistently.

## How did you test it?
Comprehensive unit test suite covering:
- Basic operation tracking
- Deletion scenarios
- Path prefix edge cases
- Operation ordering
- Deep hierarchies
- Multiple deletion points

## Potential risks
- Outputs() callers that expected non-compacted views might be affected by compaction. I couldn't find any components that will be negatively impacted.

## Documentation
No external documentation changes needed as this is an internal change.

## Is hotfix candidate?
No - this is a foundational change that should follow normal release processes.

### CODE REVIEW

File: service/history/hsm/tree.go [hunk 1]
```diff
@@ -90,17 +90,50 @@ type cachedMachine struct {
...
 
-func (o *OperationLog) IsDeleted(path []Key) bool {
-	for _, deletedPath := range o.DeletedPaths {
-		if isPathPrefix(deletedPath, path) {
-			return true
+// DeleteOperation represents the deletion of a node in the tree. When a node
+// is deleted, DeleteOperations are created for it and all its descendants.
+type DeleteOperation struct {
+	// NodePath is the full path to the deleted node
+	NodePath []Key
```
Comment (bergundy) [inline service/history/hsm/tree.go:106]: No need to export.

```suggestion
	path []Key
```

File: service/history/hsm/tree.go [hunk 2]
```diff
@@ -90,17 +90,50 @@ type cachedMachine struct {
...
+}
+
+func (d DeleteOperation) Path() []Key           { return d.NodePath }
+func (DeleteOperation) mustImplementOperation() {}
+
+// TransitionOperation represents a state transition that occurred at a specific
+// node in the tree.
+type TransitionOperation struct {
+	// NodePath is the full path to the node where the transition occurred
+	NodePath []Key
```
Comment (bergundy) [inline service/history/hsm/tree.go:116]: Same, no need to export.

File: service/history/hsm/tree.go [hunk 3]
```diff
@@ -213,29 +240,23 @@ func (n *Node) Path() []Key {
...
-	if transitions := root.opLog.TransitionsByPath[pathKey]; len(transitions) > 0 {
-		paos = append(paos, PathAndOutputs{
-			Path:    currentPath,
-			Outputs: transitions,
-		})
-	}
-
-	for _, child := range n.cache.children {
-		paos = append(paos, child.Outputs()...)
+	// Filter operations to only include those in this subtree
```
Comment (bergundy) [inline service/history/hsm/tree.go:250]: You can also check if the current node is the root (e.g. parent is nil) and skip this loop since this how this be executed by the framework.

File: service/history/hsm/tree.go [hunk 4]
```diff
@@ -386,18 +408,19 @@ func (n *Node) DeleteChild(key Key) error {
 		return err
 	}
 
-	// Mark entire subtree as deleted
+	// Mark entire subtree as deleted and record all deletions
+	root := n.root()
 	if err := child.Walk(func(n *Node) error {
 		n.cache.deleted = true
+		// Record deletion for every node in subtree
```
Comment (bergundy) [inline service/history/hsm/tree.go:415]: No need, these will be compacted anyways.

File: service/history/hsm/tree.go [hunk 5]
```diff
@@ -700,6 +725,44 @@ func (n *Node) root() *Node {
...
+			deletedPaths = append(deletedPaths, del.Path())
+		}
+	}
+
+	// Then filter transitions
+	result := make(OperationLog, 0, len(ol))
+	for _, op := range ol {
+		switch o := op.(type) {
+		case DeleteOperation:
+			// Keep all delete operations as tombstones
```
Comment (bergundy) [inline service/history/hsm/tree.go:746]: If the parent is deleted, you want to compact the child deletion away.

**Review by bergundy — COMMENTED** [review]
Didn't review the tests. Had some comments on the code. It's very close.

File: service/history/hsm/tree.go (hunk 2 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Fixed

File: service/history/hsm/tree.go (hunk 3 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Since the typical flow is for the root to be the arg, we now compact the full log and only filter if non-root. Thx.

File: service/history/hsm/tree.go (hunk 4 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Indeed! Thx

File: service/history/hsm/tree.go (hunk 5 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Done!

File: service/history/hsm/tree.go [hunk 6]
```diff
@@ -244,8 +271,7 @@ func (n *Node) Outputs() []PathAndOutputs {
 func (n *Node) ClearTransactionState() {
 	root := n.root()
 	if root.opLog != nil {
-		root.opLog.DeletedPaths = nil
-		root.opLog.TransitionsByPath = make(map[string][]TransitionOutputWithCount)
+		root.opLog = make(OperationLog, 0)
```
Comment (bergundy) [inline service/history/hsm/tree.go:274]: nit: this could be set to nil to avoid the allocation. len() and append() accept nils.

File: service/history/hsm/tree.go [hunk 7]
```diff
@@ -700,6 +730,48 @@ func (n *Node) root() *Node {
...
+func (ol OperationLog) Compact() OperationLog {
+	var deletedPaths [][]Key
+	result := make(OperationLog, 0, len(ol))
+	hasDeleteOperation := make(map[string]bool)
+
+	// First get deletion paths and record existing deletions
+	for _, op := range ol {
+		if del, ok := op.(DeleteOperation); ok {
+			deletedPaths = append(deletedPaths, del.Path())
+			pathKey := fmt.Sprintf("%v", del.Path())
```
Comment (bergundy) [inline service/history/hsm/tree.go:745]: I would probably just implement this as a tree too instead of formatting but it's not critical.

File: service/history/hsm/tree.go [hunk 8]
```diff
@@ -700,6 +730,48 @@ func (n *Node) root() *Node {
 	return root
 }
 
+// Compact filters the operation log to remove any transitions for deleted nodes while preserving deletion operations.
+// The returned log maintains deletion tombstones, but excludes transitions that occurred on deleted nodes or their
+// descendants.
+func (ol OperationLog) Compact() OperationLog {
+	var deletedPaths [][]Key
+	result := make(OperationLog, 0, len(ol))
+	hasDeleteOperation := make(map[string]bool)
```
Comment (bergundy) [inline service/history/hsm/tree.go:739]: nit: use `map[something]struct{}` for sets in Go.

File: service/history/hsm/tree.go (hunk 7 above)
Comment (bergundy) [reply service/history/hsm/tree.go:745]:
```go
type opNode struct {
	key Key
	children map[Key]opNode
	isDeleted bool
}
```

Then you can traverse the tree to know whether you're in a deleted branch.

Comment (bergundy) [reply service/history/hsm/tree.go:745]: See this for reference: https://github.com/temporalio/temporal/blob/0b9febbd0632428834589c9bdb597deaaf26648b/tools/testrunner/testrunner.go#L44

File: service/history/hsm/tree.go (hunk 6 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Thx for the tip! Done

File: service/history/hsm/tree.go (hunk 8 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Thx! Done

File: service/history/hsm/tree.go (hunk 7 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: I _think_ this is what you had in mind. Please take a look.

File: service/history/hsm/tree.go [hunk 9]
```diff
@@ -700,10 +727,85 @@ func (n *Node) root() *Node {
...
+		for _, op := range n.operations {
+			if _, ok := op.(DeleteOperation); ok {
+				result = append(result, op)
+			}
+		}
+	} else {
+		result = append(result, n.operations...)
+	}
+
+	for _, child := range n.children {
```
Comment (bergundy) [inline service/history/hsm/tree.go:807]: You should avoid adding operations for children if the node is already deleted. That's the point of compaction.

Comment (justinp-tt) [reply service/history/hsm/tree.go:805]: We actually do avoid adding non-delete operations for children if the node in question is already deleted (proven by test cases). Notice that we check for "isDeleted" and, if true, we only add the DeleteOperation.

That being said, we can make this a bit cleaner. Updated!

Breakdown for clarity:

1. Tree Construction (Compact function):
  * Creates empty tree with root node
  * Take each op from log (in order) and insert into tree at proper path location

2. Operation Insertion (insert method):
  * If we're at the target path:
    * If deleted
      * mark node deleted
      * replace all ops with singular delete op
      * create cascade delete ops for all descendants
    * If not deleted
      * add op to node's op list
  * If we're not at target path:
    * Create/get next child node in path
    * Recursively inserts op with remaining path

3. Collection (collect):
* Depth-first traversal 
* Collects ops from each node in order (parent then children)
**Note: deleted nodes only contain delete ops since insertion replaced other ops**

File: service/history/hsm/tree.go [hunk 10]
```diff
@@ -700,10 +727,85 @@ func (n *Node) root() *Node {
...
+	}
+}
+
+// insert adds an operation at the specified path in the tree.
+func (n *opNode) insert(path []Key, op Operation) {
+	if len(path) == 0 {
+		if deleteOp, ok := op.(DeleteOperation); ok {
+			n.isDeleted = true
+			for k, child := range n.children {
+				child.insert(nil, DeleteOperation{path: append(slices.Clone(deleteOp.path), k)})
```
[the remark this answers is not in the document]
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Done!

File: service/history/hsm/tree.go [hunk 11]
```diff
@@ -700,10 +727,85 @@ func (n *Node) root() *Node {
...
+	if len(ol) == 0 {
+		return ol
+	}
+
+	root := newOpNode(Key{})
+	for _, op := range ol {
+		root.insert(op.Path(), op)
+	}
+
+	return root.collect()
```
[the remark this answers is not in the document]
Comment (justinp-tt) [reply service/history/hsm/tree.go:743]: Fixed

File: service/history/hsm/tree.go [hunk 12]
```diff
@@ -90,17 +96,90 @@ type cachedMachine struct {
...
+// String returns a human-readable representation of a path
+func formatPath(path []Key) string {
+	if len(path) == 0 {
+		return "<root>"
+	}
+	var parts []string
+	for _, key := range path {
+		parts = append(parts, key.String())
+	}
+	return strings.Join(parts, "/")
```
Comment (bergundy) [inline service/history/hsm/tree.go:108]: This would be ambiguous if the keys contained slashes but it's fine, it's just for debugging.

File: service/history/hsm/tree.go [hunk 13]
```diff
@@ -90,17 +96,90 @@ type cachedMachine struct {
...
+	Path() []Key
+	mustImplementOperation()
 }
 
-func (o *OperationLog) IsDeleted(path []Key) bool {
-	for _, deletedPath := range o.DeletedPaths {
-		if isPathPrefix(deletedPath, path) {
-			return true
+// DeleteOperation represents the deletion of a node in the tree. When a node
+// is deleted, DeleteOperations are created for it and all its descendants.
```
Comment (bergundy) [inline service/history/hsm/tree.go:121]:
```suggestion
// DeleteOperation represents the deletion of a node in the tree.
```

File: service/history/hsm/tree.go [hunk 14]
```diff
@@ -90,17 +96,90 @@ type cachedMachine struct {
...
+type DeleteOperation struct {
+	// NodePath is the full path to the deleted node
```
Comment (bergundy) [inline service/history/hsm/tree.go:123]: Can you add punctuation to all docstrings please?

File: service/history/hsm/tree.go [hunk 15]
```diff
@@ -90,17 +96,90 @@ type cachedMachine struct {
...
+			// Fallback for unknown operation types
+			ops = append(ops, fmt.Sprintf("%T(%v)", op, op.Path()))
+		}
+	}
+	return fmt.Sprintf("[%s]", strings.Join(ops, ", "))
+}
+
+// IsDeleted returns true if the given path or any of its ancestors has been deleted. This is used to determine if
+// operations are valid for a given path.
+func (ol OperationLog) IsDeleted(path []Key) bool {
```
Comment (bergundy) [inline service/history/hsm/tree.go:177]: Not in use. You can delete this.

File: service/history/hsm/tree.go [hunk 16]
```diff
@@ -375,7 +432,8 @@ func (n *Node) AddChild(key Key, data any) (*Node, error) {
 	return node, nil
 }
 
-// DeleteChild deletes an immediate child node and all its descendants from the tree.
+// DeleteChild marks a child node and all its descendants as deleted, removing them from the cache. After deletion,
+// nodes will only see operations relevant to their subtree via prefix matching of paths.
```
Comment (bergundy) [inline service/history/hsm/tree.go:436]: This comment is unclear to me. WDYM by nodes will only see operations?
I'd keep this comment geared towards users of the (soon to be deprecated) framework.

File: service/history/hsm/tree.go [hunk 17]
```diff
@@ -700,10 +761,116 @@ func (n *Node) root() *Node {
...
+		if len(path) == 0 {
+			if n.isDeleted {
+				if deleteOp, ok := op.(DeleteOperation); ok {
+					result = append(result, deleteOp)
+				}
+			} else {
+				result = append(result, op)
+			}
+			continue
+		}
```
Comment (bergundy) [inline service/history/hsm/tree.go:840]: Seems like there should be a way to merge this with the `Check "target" node` code below.

File: service/history/hsm/tree_test.go [hunk 18]
```diff
@@ -543,13 +563,30 @@ func TestOperationLog_IsDeleted(t *testing.T) {
...
+	require.NoError(t, err)
 
-	// Non-deleted sibling should still have outputs
-	siblingOutputs := l2_sibling.Outputs()
-	require.Len(t, siblingOutputs, 1)
-	require.Equal(t, l2_sibling.Path(), siblingOutputs[0].Path)
+	var foundDelete bool
+	for _, op := range opLog {
+		if delOp, ok := op.(hsm.DeleteOperation); ok {
+			if reflect.DeepEqual(delOp.Path(), l2.Path()) {
```
Comment (bergundy) [inline service/history/hsm/tree_test.go:572]: `slices.Equal` also works here.

Comment (bergundy) [reply service/history/hsm/tree_test.go:572]: Or replace this whole loop with https://pkg.go.dev/slices#ContainsFunc.

File: service/history/hsm/tree_test.go [hunk 19]
```diff
@@ -717,11 +796,214 @@ func TestNode_MultipleDeletedPaths(t *testing.T) {
...
+	// Add transitions to all nodes
+	for _, node := range []*hsm.Node{node1, node10, node11, child1, deepNode1} {
+		err = hsm.MachineTransition(node, func(d *hsmtest.Data) (hsm.TransitionOutput, error) {
+			d.SetState(hsmtest.State2)
+			return hsm.TransitionOutput{}, nil
+		})
+		require.NoError(t, err)
+	}
+
+	err = root.DeleteChild(hsm.Key{Type: def1.Type(), ID: "node1"})
```
Comment (bergundy) [inline service/history/hsm/tree_test.go:852]: nit: `node1.Key`.

**Review by bergundy — APPROVED** [review]
Looks great. Nothing blocking functionally. Feel free to address whatever and merge.

File: service/history/hsm/tree.go (hunk 14 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Done

File: service/history/hsm/tree.go (hunk 16 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go:435]: // DeleteChild marks a child node and all its descendants as deleted, removing them from the cache. No transitions will
// be allowed after deleting a child.

File: service/history/hsm/tree.go (hunk 15 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Removed

File: service/history/hsm/tree_test.go (hunk 18 above)
Comment (justinp-tt) [reply service/history/hsm/tree_test.go]: Thx for the tip!

File: service/history/hsm/tree.go (hunk 17 above)
Comment (justinp-tt) [reply service/history/hsm/tree.go]: Merged!

File: service/history/hsm/tree_test.go (hunk 19 above)
Comment (justinp-tt) [reply service/history/hsm/tree_test.go:843]: Are you saying you would prefer the `ID` to be `"node1.Key"`?
